opwindow.go

```diff
@@ -0,0 +1,220 @@
+	cond := sync.NewCond(&sync.Mutex{})
+	myctx, can := context.WithCancel(ctx)
+	q := &OpWindow{
+		cond:       cond,
+		ctx:        myctx,
+		can:        can,
+		depth:      depth,
+		width:      width,
+		q:          list.New(),
+		entries:    map[ID]*OpSet{},
+		windowedBy: windowedBy.Nanoseconds(),
+	}
+	go func() {
+		t := time.NewTicker(250 * time.Millisecond)
+		for {
+			select {
+			case <-q.ctx.Done():
+				return
+			case n := <-t.C:
+				// signal someone to wake up incase we have any Items falling
+				// out of the window.
+				q.cond.L.Lock()
+				if q.itemsReady(n) {
+					q.cond.Signal()
+				}
+				q.cond.L.Unlock()
+			}
+		}
+	}()
+	return q
+}
+
+// Close releases resources associated with this callgroup, by canceling
+// the context. The owner of this OpWindow should either call Close or cancel
+// the context, both are equivalent.
+func (q *OpWindow) Close() {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+
+	q.can()
+	// alert all dequeue calls that they should wake up and return.
+	q.windowedBy = 0 // turn off windowing so everything is dequeue
+	q.cond.Broadcast()
+	return
+}
+
+// Len returns the number of uniq IDs in the queue, that is the depth of the
+// queue.
+func (q *OpWindow) Len() int {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	return q.q.Len()
+}
+
+// Enqueue add the op to the queue.  If the ID already exists then the Op
+// is added to the existing OpSet for this ID, otherwise it's inserted as
+// a new OpSet.
+//
+// Enqueue doesn't block if the queue if full, instead it returns a
+// ErrQueueSaturated error.
+func (q *OpWindow) Enqueue(id ID, op *Op) error {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+
+	if q.q.Len() >= q.depth {
+		return ErrQueueSaturatedDepth
+	}
+
+	set, ok := q.entries[id]
+	if !ok {
+		set = newOpSet()
+		// This is a new item, so we need to insert it into the queue.
+		q.enqueue(id)
+
+		// Signal one waiting go routine to wake up and Dequeue
+		// I believe we only need to signal if we enqueue a new item.
+		// Consider the following possible states the queue could be in  :
+		//   1. if no one is currently waiting in Dequeue, the signal isn't
+		//      needed and all items will be dequeued on the next call to
+		//      Dequeue.
+		//   2. One or Many go-routines are waiting in Dequeue because it's
+		//      empty, and calling Signal will wake up one.  Which will
+		//      dequeue the item and return.
+		//   3. At most One go-routine is in the act of Dequeueing existing
+		//      items from the queue (i.e. only one can have the lock and be
+		//      in the "if OK" condition within the forloop in Dequeue).  In
+		//      which cause the signal is ignored and after returning we
+		//      return to condition (1) above.
+		// Note signaled waiting go-routines will not be able the acquire
+		// the condition lock until this method call returns, finishing
+		// its append of the new operation.
+		q.cond.Signal()
+	}
+
+	if len(set.Ops()) >= q.width {
```

Comment:
Do you want to move this to line 104? Otherwise you can enqueue the id but not update the `entries` map

Reply:
The only way the `set` length could be greater than width is if the set isn't new, in which case it's already enqueued in the list and entered into the map.   I do agree this looks weird and the code on 136 is redundant since it inserts the same set over and over... It could safe a tiny bit of performance if we moved `q.entries[id] = set` into the `if !ok {` block above.

I'm going to investigate this and try to make the code more readable.